fluent/resolver.py

```diff
@@ -0,0 +1,391 @@
+from __future__ import absolute_import, unicode_literals
+
+from datetime import date, datetime
+from decimal import Decimal
+
+import attr
+import six
+
+from .exceptions import FluentCyclicReferenceError, FluentReferenceError
+from .syntax.ast import (AttributeExpression, CallExpression, Message, MessageReference, NumberLiteral,
+                         Pattern, Placeable, SelectExpression, StringLiteral, Term, TermReference, TextElement,
+                         VariableReference, VariantExpression, VariantList, Identifier)
+from .types import FluentDateType, FluentNone, FluentNumber, fluent_date, fluent_number
+from .utils import numeric_to_native
+
+try:
+    from functools import singledispatch
+except ImportError:
+    # Python < 3.4
+    from singledispatch import singledispatch
+
+
+text_type = six.text_type
+
+# Prevent expansion of too long placeables, for memory DOS protection
+MAX_PART_LENGTH = 2500
+
+# Prevent messages with too many sub parts, for CPI DOS protection
+MAX_PARTS = 1000
+
+
+# Unicode bidi isolation characters.
+FSI = "\u2068"
+PDI = "\u2069"
+
+
+@attr.s
+class ResolverEnvironment(object):
+    context = attr.ib()
+    args = attr.ib()
+    errors = attr.ib()
+    dirty = attr.ib(factory=set)
+    part_count = attr.ib(default=0)
+
+
+def resolve(context, message, args, errors=None):
```

Comment:
After a while away, I have my own comment here!

Instead of passing `context` to resolve, I think it would probably be nicer if we passed the things from `MessageContext` that are actually needed, namely:
* the message and term dicts, possibly combined into a single dict.
* locale
* use_isolating
* functions

This would be a bit more verbose, but it would have the effect that the resolver code no longer needs to access internals of `MessageContext` (e.g. `._babel_locale`). Also `plural_form_for_number` would then come off `MessageContext` (which isn't a great home for it) and go somewhere better.

It would also make this code similar to the equivalent code in the [compiler implementation](https://github.com/projectfluent/python-fluent/pull/71/files#diff-f1879dcf9535775ee300cd035561340dR153)

Reply:
I think that passing a `context` is quite safe here and I saw Stas using that convention around

I do agree that a single dict for all IDs would be useful, but instead of passing a locale, I'd recommend passing a locale fallback chain as provided to the `FluentBundle` constructor.

Reply:
@zbraniecki - the problem with passing a locale fallback chain is that `resolve` would then have to convert this to a Babel locale object every time it is called, in other words for every single call to `format` (even though it will always return the same thing), rather than doing it once up front.